banyan/resource_service.go

```diff
@@ -1659,3 +1595,256 @@ func resourceServiceDelete(ctx context.Context, d *schema.ResourceData, m interf
+	v := make(map[string]interface{})
+	port, err := strconv.Atoi(toFlatten.Port) // need to convert this to int
+	if err != nil {
+		diagnostics = diag.Errorf("Could not convert BackendTarget.spec.backend.target.port int to string %v", toFlatten.Port)
+	}
+	v["client_certificate"] = toFlatten.ClientCertificate
+	v["name"] = toFlatten.Name
+	v["tls"] = toFlatten.TLS                  // might need to convert this to string
+	v["tls_insecure"] = toFlatten.TLSInsecure // might need to convert this to string
+	v["port"] = port
+	flattened = append(flattened, v)
+	return
+}
+
+func flattenServiceCertSettings(toFlatten service.CertSettings) (flattened []interface{}) {
+	v := make(map[string]interface{})
+	v["custom_tls_cert"] = flattenServiceCustomTLSCert(toFlatten.CustomTLSCert)
+	v["dns_names"] = toFlatten.DNSNames
+	v["letsencrypt"] = toFlatten.LetsEncrypt
+
+	flattened = append(flattened, v)
+	return
+}
+
+func flattenServiceCustomTLSCert(toFlatten service.CustomTLSCert) (flattened []interface{}) {
+	v := make(map[string]interface{})
+	v["cert_file"] = toFlatten.CertFile
+	v["enabled"] = toFlatten.Enabled
+	v["key_file"] = toFlatten.KeyFile
+	flattened = append(flattened, v)
+	return
+}
+
+func flattenServiceHTTPSettings(toFlatten service.HTTPSettings) (flattened []interface{}) {
+	v := make(map[string]interface{})
+	v["enabled"] = toFlatten.Enabled
+	v["exempted_paths"] = flattenServiceExemptedPaths(toFlatten.ExemptedPaths)
+	v["http_health_check"] = flattenServiceHTTPHealthCheck(toFlatten.HTTPHealthCheck)
+	v["http_redirect"] = flattenServiceHTTPRedirect(toFlatten.HTTPRedirect)
+	v["headers"] = toFlatten.Headers
+	v["oidc_settings"] = flattenServiceOIDCSettings(toFlatten.OIDCSettings)
+	flattened = append(flattened, v)
+	return
+}
+
+func flattenServiceExemptedPaths(toFlatten service.ExemptedPaths) (flattened []interface{}) {
+	v := make(map[string]interface{})
+	v["enabled"] = toFlatten.Enabled
+	v["paths"] = toFlatten.Paths
+	v["pattern"] = flattenServicePatterns(toFlatten.Patterns)
+	flattened = append(flattened, v)
+	return
+}
+
+func flattenServicePatterns(toFlatten []service.Pattern) (flattened []interface{}) {
+	flattened = make([]interface{}, len(toFlatten), len(toFlatten))
+
+	for idx, item := range toFlatten {
+		v := make(map[string]interface{})
+		v["hosts"] = flattenServiceHosts(item.Hosts)
+		v["mandatory_headers"] = item.MandatoryHeaders
+		v["methods"] = item.Methods
+		v["paths"] = item.Paths
+		v["source_cidrs"] = item.SourceCIDRs
+		flattened[idx] = v
+	}
+	return
+}
+
+func flattenServiceHosts(toFlatten []service.Host) (flattened []interface{}) {
+	flattened = make([]interface{}, len(toFlatten), len(toFlatten))
+
+	for idx, item := range toFlatten {
+		v := make(map[string]interface{})
+		v["origin_header"] = item.OriginHeader
+		v["target"] = item.Target
+		flattened[idx] = v
+	}
+	return
+}
+
+func flattenServiceHTTPHealthCheck(toFlatten service.HTTPHealthCheck) (flattened []interface{}) {
+	v := make(map[string]interface{})
```

Comment:
Why is `flattened` an array? There's only ever one element.

Also, I think this style looks a little nicer:

```
v := map[string]interface{}{
    "addresses": toFlatten.Addresses,
    "enabled": toFlatten.Enabled,
...
}

flattened = []interface{}{v}
return
```

Reply:
I guess you're sticking to a single pattern for all the flatten functions. Seems ok.